### ISSUE TITLE
Question: Setting up link states for models with more than 2 states per cell + more

### ISSUE THREAD

**Issue by alderonruiner** [landlab/landlab#1247]
Hi, I'm currently trying to determine if this tool is suitable for a project I'm working on involving the fire spread prediction with heat camera feedback. I have several questions:

1. Are there any examples for setting up link-states for a CA that has 3 or more states per cell? I attempted to modify the isotropic example to create some form of fire-spread model. I am unsure if it is remotely close to being set up correctly. Are the cell transitions functional transitions of other cells?

2. I have up to this point been attempting to build my own simulation in Matlab using generalised tensor products and Markov chains. Is this also what is happening under the hood in landlab for CA functionality?

3. If I can model fire spread with this tool, would it be possible to add feedback from a source to reduce prediction error every so many time steps? Or because of the CT nature is this not possible and a discrete time model would be required?

**Comment by kbarnhart** [comment]
@alderonruiner I'm not super familiar with the CA submodule (@gregtucker is the go to person). The two places I'd look to understand its functionality are:

https://landlab.readthedocs.io/en/master/user_guide/index.html#celllab-cts-user-guide
https://landlab.readthedocs.io/en/master/reference/ca/index.html

**Comment by gregtucker** [comment]
@alderonruiner thanks for your interest in CellLab-CTS. When you say 3 states per cell, I'm not sure whether you mean a cell can take one of 3 possible states (which the package can definitely do), or a cell has a state in each of 3 different categories. If it's the latter, then the basic answer is no. But you could potentially get around this by creating states as combinations; so for example if you have burning vs. not burning and trees vs. grass, you could create four states of burning grass, burning trees, not-burning grass, and not-burning trees.

Under the hood, the package works by identifying pairs of adjacent cells. For each pair, there is a certain probability per time (possibly zero) of transition to a different pair. The code uses Poisson process theory to find the time that this transition occurs by drawing from an exponential waiting-time distribution. The transition is scheduled in a priority queue, and the program iterates through the events in the queue. When one or both cells associated with a future transition change state, the scheduled transition is invalidated. See this paper for more: https://gmd.copernicus.org/articles/9/823/2016/

You could probably engineer a way to do periodic updates to the cell states. The model is discrete in space but continuous in time. The package is engineered so that you can run it for a specified duration, then pause to do something to it; for example, in these two papers this method is used to periodically inject tectonic uplift:

https://esurf.copernicus.org/articles/6/563/2018/

https://doi.org/10.1029/2019JF005305

**Comment by alderonruiner** [comment]
@gregtucker yep, a cell that can take on one of three possible states.

I ended up using a naïve approach, in MATLAB, to complete my thesis, which ended up coming back to bite me. I've recently wanted to expand my model to somehow factor in elevation and see what the results might look like, I managed to include wind direction and fire spotting as well as some very basic overall parameters (general density, moisture level etc.). I know longer have access to MATLAB so I began converting it to Python, which has displayed the deficiencies in the original coding approach, as well as the pain it has been to find similar Python functions.

I remembered landlab and have been reexploring it. I'll aim just to set up a very basic cellular automata first using three states:

- F — Fuel
- B — Burning
- O — Out

From your explanation and my vague memory of the the papers, a pairwise state is just two cells next to each other, such as if cell 1 was F and cell 2 was B, then pairwise transition would result in another pairwise state (cell 1 B and cell 2 B)? My basic rules for the cellular automata model are described in Section 4.1 of https://www.researchgate.net/publication/352150208_Fire_Spread_Prediction

These are the pairwise transitions I came up with for such a model:

        State 0 - Fuel (F)
        State 1 - Burning (B)
        State 2 - Out (O)

        Pair state      Transition to       Process 
        ==========      =============       ======= 
        0 (0-0)         (none)              -
        1 (0-1)         2 (1-0)             ignition 
        2 (1-0)         1 (0-1)             ignition
        3 (1-1)         7 (2-2)             extinguish
        4 (1-2)         7 (2-2)             extinguish
        5 (2-0)         (none)              -
        6 (2-1)         7(2-2)              extinguish
        7 (2-2)         (none)              -    

If you have any insights/suggestions, I'd appreciate to hear them, otherwise this issue can be closed.
